flow/aggregates.py

```diff
@@ -538,6 +538,19 @@ def _register_aggregates(self, project):
         """
         self._project = project
 
+    def __setstate__(self, data):
```

Comment:
I ran some tests and wanted to update my question. I don't think this block of code is needed if we're using `cloudpickle`, since the only problem I'm aware of is an issue with pickling locally defined functions (which cloudpickle solves). If that's correct, we shouldn't need to manually define `__getstate__` and `__setstate__` for any classes.

@b-butler @kidrahahjo Is this correct? If not, could you help provide a minimal case where these are required and commit it to the tests?

Reply:
The project attribute gets lost if we don't get them manually. Not sure why is this happening though.
The error can be reproduced by executing this `pytest tests/test_project -k parallel -x` when these `__getstate__` and `__setstate__` methods are removed.

Reply:
I spent quite a bit of time and determined the core issue. The problem has to do with pickling/unpickling the project configuration. We monkey-patch the signac project configuration in signac-flow and when that gets unpickled, it points to the `configobj` from signac. Since the `flow` key isn't defined, it can't re-construct the project configuration during unpickling.

Reply:
I think we can adopt this workaround with `__getstate__` / `__setstate__` for now, and add a comment describing the core issue. I'll add that to this PR and approve it.

Reply:
This gets even better. I got deeper into the problem, and found a cleaner solution. Here's my best explanation:

Recall that the FlowProject owns aggregate stores, which may include a `_DefaultAggregateStore` that holds a reference to the original FlowProject. The circular reference is not intrinsically a problem, but there is a complication. The `_DefaultAggregateStore` is used as a key in a dict, which means that it must be hashable. When the `__hash__` method is called during unpickling, the `_DefaultAggregateStore`'s hash function returns `hash(repr(self._project))`. This means that in order to reconstruct the pickled project, the FlowProject (because of the `_DefaultAggregateStore`) must compute its own `__repr__`. However, that relies on `FlowProject._config` already being set, since the `repr` returns something like `Project.get_project('/path/to/project')` via the `Project._rd()` method, which requires the root directory from the config.

If it were possible to control the order in which data was unpickled, it might be possible to break this circular chain and avoid the error. However, rather than defining `__getstate__` and `__setstate__` for the `FlowProject` or `_DefaultAggregateStore`, I think it would be easiest just to store `self._hash_repr = repr(self._project)` when the `_DefaultAggregateStore` is created, and define it `__hash__` method to return `hash(self._hash_repr)`. It ends up being only 2 lines (instead of adding entirely new methods) and I think it might be safer and more future-proof.

Reply:
@kidrahahjo @b-butler Tagging you both so you can see the above comment. I think this PR is good to go, now that I've fixed the issue described above. Is my explanation clear? Did I leave a good enough comment in the new code?